connect/src/state/context.rs

```diff
@@ -100,26 +130,35 @@ impl ConnectState {
             }
         }
 
+        self.fill_up_context = ContextType::Default;
+        self.set_active_context(ContextType::Default);
         self.update_restrictions()
     }
 
-    pub fn get_context_uri_from_context(context: &Context) -> Option<&String> {
-        if !context.uri.starts_with(SEARCH_IDENTIFIER) {
-            return Some(&context.uri);
-        }
+    pub fn valid_resolve_uri(uri: &str) -> Option<&str> {
+        (!uri.starts_with(SEARCH_IDENTIFIER)).then_some(uri)
+    }
 
-        context
-            .pages
-            .first()
-            .and_then(|p| p.tracks.first().map(|t| &t.uri))
+    pub fn get_context_uri_from_context(context: &Context) -> Option<&str> {
+        match Self::valid_resolve_uri(&context.uri) {
+            Some(uri) => Some(uri),
+            None => context
+                .pages
+                .first()
+                .and_then(|p| p.tracks.first().map(|t| t.uri.as_ref())),
+        }
     }
 
     pub fn set_active_context(&mut self, new_context: ContextType) {
         self.active_context = new_context;
 
-        let ctx = match self.get_context(&new_context) {
+        let player = self.player_mut();
+        player.context_metadata.clear();
+        player.restrictions.clear();
```

Comment:
How much can `context_metadata` and `restrictions` differ in size? Clearing keeps the memory allocated; does it make sense to free it (shrink it) or keep it around? I can imagine it makes sense to keep it, so just asking.

Reply:
Never thought about that before. Anyhow, the metadata provided for context can be a big difference. For example an artists context is about two entries and the discovery weekly has more then 10 entries. For restrictions it's usually one to six, but it could be more.

For restrictions it probably doesn't make much sense to free it, but for the metadata could be. But could also be that for just ten entries it isn't worth it to free the instance.

Btw. do you mean by freeing dropping the current map and creating a new instance?

Reply:
Two options:

1. Dropping then assigning a new object, like you say, or

2. Calling `shrink_to_fit`. 

Honestly I don't know about the performance difference between dropping+creating versus clearing+shrinking, probably both just fine in this case.